warehouse/templates/manage/account/totp-provision.html

```diff
@@ -0,0 +1,67 @@
+{#
+ # Licensed under the Apache License, Version 2.0 (the "License");
+ # you may not use this file except in compliance with the License.
+ # You may obtain a copy of the License at
+ #
+ # http://www.apache.org/licenses/LICENSE-2.0
+ #
+ # Unless required by applicable law or agreed to in writing, software
+ # distributed under the License is distributed on an "AS IS" BASIS,
+ # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ # See the License for the specific language governing permissions and
+ # limitations under the License.
+-#}
+{% extends "manage/manage_base.html" %}
+
+{% set user = request.user %}
+{% set title = "Set up two factor authentication using an application (TOTP)" %}
+
+{% block title %}{{ title }}{% endblock %}
+
+{# Hide mobile search on manager pages #}
+{% block mobile_search %}{% endblock %}
+
+{% block main %}
+<h1 class="page-title">{{ title }}</h1>
+<section>
+  <h2>How 2FA with TOTP works</h2>
+
+  <ol>
+    <li>When you sign into PyPI, you will be asked to provide your username and password, as usual</li>
+    <li>You will then see a screen asking for an additional authentication code</li>
+    <li>Use the application of your choice (provisioned below) to generate a code</li>
+    <li>Enter the code to access your PyPI account</li>
+  </ol>
+
+  <h2>Provisioning your application</h2>
+
+  <p>You can provision any compatible TOTP application to generate two factor authentication codes for PyPI. We recommend <a href="https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2&hl=en_GB">Google Authenticator for Android</a> or <a href="https://itunes.apple.com/gb/app/google-authenticator/id388497605?mt=8">iPhone</a>, <a href="https://support.1password.com/one-time-passwords/">1Password</a>, <a href="https://authy.com/">Authy</a> or <a href="https://lastpass.com/auth/">LastPass Authenticator</a>.</p>
```

Comment:
May I ask adding FreeOTP/FreeOTP+ to the list? The reason for this is that they are open source and others are AFAIK proprietary which makes them questionable from the security recommendations PoV.

Reply:
FreeOTP sounds good to me! I personally used FreeOTP+ back when I was on Android, but in retrospect I'm not actually sure about recommending it -- it provides a TOTP export functionality that compromises the application's ability to behave as a true second factor.

Reply:
Well, you can as well feed the same seed to several apps/devices and you'll get the same issue. I don't consider export a truly bad thing as it provides a backup way other than recovery codes.
If you know anything other which is open source, that would probably work as well.
But my point is that trusting proprietary software is probably more of a security risk vs. having an export feature.
BTW FreeOTP+ also uses more of nice Android APIs to protect data, like hiding previews on the app switch screen unlike Google Authenticator and others.

Reply:
> Well, you can as well feed the same seed to several apps/devices and you'll get the same issue.

Yep, that's why we settled on the warning right below the recommended application list:

```
For security reasons, you should only provision one device/application at a time.
```

TOTP's symmetricity makes it hard for us to *enforce* good client security decisions, so I think we should make an effort to *suggest* the strongest defaults. That, IMO, includes not suggesting TOTP applications that weaken their status as a second factor by allowing users to extract TOTP secrets. It might also be a good reason to remove 1Password from that list, since also violates the second factor property.

FWIW, I *completely* agree that it's not a truly bad thing -- for expert users who understand that exports violate the second factor property. That's why I used it, back when I did! But I don't think we should encourage that as one of the default recommendations, for essentially the same reason that we shouldn't require recovery codes.

FreeOTP is the open source TOTP application that I'm most familiar with. I see that [andOTP](https://github.com/andOTP/andOTP) appears to be popular, although I don't know anything about it in terms of history/behavior. Another search on F-Droid brings up [otp-authenticator](https://github.com/0xbb/otp-authenticator). I agree that trusting a proprietary TOTP application is a security hazard, although I think applications that encourage lay-users to back up their TOTP secrets are more of a direct security risk.

Reply:
So maybe just add another warning about that? I mean like "hey, there's other apps which you can actually verify if you're truly concerned about security"

Reply:
Yeah that sounds good to me. I think we could integrate it directly into the recommendations, e.g. "We recommend FreeOTP, ... as open source TOTP applications. Google Authenticator, ..., are also available as proprietary alternatives." Thoughts @nlhkabu?